source/httpfs/http_filesystem.go

```diff
@@ -0,0 +1,145 @@
+package httpfs
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"sort"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// Config object.
+// Dir - base directory for serving files
+type Config struct {
+	Dir string
+}
+
+// HTTPFS is migration source that uses http.FileSystem as the source of
+// migration scripts.
+type HTTPFS struct {
+	path       string
+	fs         http.FileSystem
+	migrations *source.Migrations
+}
+
+// WithInstance creates HTTPFS instance that uses provided FileSystem to source
+// files. config argument can be used to select a subdirectory as the root for
+// sourcing all files.
+func WithInstance(fs http.FileSystem, config *Config) (source.Driver, error) {
+	dir, err := fs.Open(config.Dir)
+	if err != nil {
+		return nil, fmt.Errorf("can't open directory with migrations: %s", err)
+	}
+
+	files, err := dir.Readdir(-1)
+	dir.Close()
+	if err != nil {
+		return nil, fmt.Errorf("can't read files in migrations directory: %s", err)
+	}
+
+	sort.Slice(files, func(i, j int) bool { return files[i].Name() < files[j].Name() })
+
+	s := &HTTPFS{
+		path:       config.Dir,
+		fs:         fs,
+		migrations: source.NewMigrations(),
+	}
+
+	for _, fi := range files {
+		if fi.IsDir() {
+			continue
+		}
+
+		m, err := source.DefaultParse(fi.Name())
+		if err != nil {
+			continue // ignore files that we can't parse
+		}
+
+		if !s.migrations.Append(m) {
+			return nil, fmt.Errorf("unable to parse file %v", fi.Name())
+		}
+	}
+
+	return s, nil
+}
+
+// Open is not implemented as many FileSystems cannot be opened or have
+// custom initialization.
+//
+// Common use:
+//   plainFileSystem, err := WithInstance(http.Dir(tmpDir), &Config{})
+//
+//   escEmbeddedFileSystem, err := WithInstance(FS(false), &Config{})
+func (f *HTTPFS) Open(url string) (source.Driver, error) {
```

Comment:
Needs to be removed as mentioned [here](https://github.com/golang-migrate/migrate/pull/163#discussion_r252909158)